Add native formatting of AUR info search

ChuckDaniels87 opened the pull request:
The search of AUR information is now formatted in `pacaur`. Information is still fetched using `cower` but the style is closer to `pacman -Si` output.

I've tested this code in LANG=en_US.UTF-8 and LANG=es_ES.UTF-8 and it seems to work.

_Edit: related to #329_

rmarquis commented on pacaur:
```diff
@@ -1103,9 +1103,70 @@ SearchInfoAur() {
+    format_strings=( $"Name" $"Version" $"Description" $"URL" $"AUR Page" $"Licenses" $"Provides" $"Depends on" $"Make Deps" $"Optional Deps" $"Conflicts With" $"Replaces" $"Mantainer" $"Votes" $"Popularity" $"Out of Date" $"Submitted" $"Last Modified" ) 
+    
+    # in some languages more tabs are needed to properly adjust the info
+    format_tab=16
```
How did you come with that number? If that value is different from one language to another one, it should be ideally retrieved dynamically.

ChuckDaniels87 replied:
I thought about it, but it doesn't seem to follow any kind of rule. At first, I thought they were multiple of 8 but they are not. In English is 16 and in Spanish is 26. We could try more languages or analyse the `pacman` code to figure out if there is any rule at all.

ChuckDaniels87 replied:
Summary (taken from `pacman` format):

* **en_US.UTF-8 :** 16 
* __es_ES.UTF-8 :__ 26
* __it_IT.UTF-8 :__ 26
* __pl_PL.UTF-8 :__ 22
* __fr_FR.UTF-8 :__ 25
* __pt_PT.UTF-8 :__ 32

I can't see any rule checking the output. I think the most logical approach is to read LANG environmental variable and manually set `format_tab` for all the languages supported by `pacaur`.

rmarquis replied:
No, the implementation should be generic here. No hardcoding.

ChuckDaniels87 replied:
I've found the code in `pacman` that is in charge of formatting the info of the packages. It seems to set the titles from the longest string in the locale. In the case of Spanish lang, is a string that does not appear in the cower recovered info. In the case of Portuguese, `pacman` output seems broken (try `LANG=pt_PT.UTF-8 pacman -Si gnome-shell`).

Anyways, we could use their approach to solve the problem taking the longest title for each locale (I've tried that and it seems to work). The output won't have the same indentation of pacman output but it would be coherent.

`src/pacman/package.c` from line **77** to **145**.
```c
/* As of 2015/10/20, the longest title (all locales considered) was less than 30
 * characters long. We set the title maximum length to 50 to allow for some
 * potential growth.
 */
#define TITLE_MAXLEN 50

static char titles[_T_MAX][TITLE_MAXLEN * sizeof(wchar_t)];

/** Build the `titles` array of localized titles and pad them with spaces so
 * that they align with the longest title. Storage for strings is stack
 * allocated and naively truncated to TITLE_MAXLEN characters.
 */
static void make_aligned_titles(void)
{
	unsigned int i;
	size_t maxlen = 0;
	int maxcol = 0;
	static const wchar_t title_suffix[] = L" :";
	wchar_t wbuf[ARRAYSIZE(titles)][TITLE_MAXLEN + ARRAYSIZE(title_suffix)];
	size_t wlen[ARRAYSIZE(wbuf)];
	int wcol[ARRAYSIZE(wbuf)];
	char *buf[ARRAYSIZE(wbuf)];
	buf[T_ARCHITECTURE] = _("Architecture");
	buf[T_BACKUP_FILES] = _("Backup Files");
	buf[T_BUILD_DATE] = _("Build Date");
	buf[T_COMPRESSED_SIZE] = _("Compressed Size");
	buf[T_CONFLICTS_WITH] = _("Conflicts With");
	buf[T_DEPENDS_ON] = _("Depends On");
	buf[T_DESCRIPTION] = _("Description");
	buf[T_DOWNLOAD_SIZE] = _("Download Size");
	buf[T_GROUPS] = _("Groups");
	buf[T_INSTALL_DATE] = _("Install Date");
	buf[T_INSTALL_REASON] = _("Install Reason");
	buf[T_INSTALL_SCRIPT] = _("Install Script");
	buf[T_INSTALLED_SIZE] = _("Installed Size");
	buf[T_LICENSES] = _("Licenses");
	buf[T_MD5_SUM] = _("MD5 Sum");
	buf[T_NAME] = _("Name");
	buf[T_OPTIONAL_DEPS] = _("Optional Deps");
	buf[T_OPTIONAL_FOR] = _("Optional For");
	buf[T_PACKAGER] = _("Packager");
	buf[T_PROVIDES] = _("Provides");
	buf[T_REPLACES] = _("Replaces");
	buf[T_REPOSITORY] = _("Repository"); :
	buf[T_REQUIRED_BY] = _("Required By");
	buf[T_SHA_256_SUM] = _("SHA-256 Sum");
	buf[T_SIGNATURES] = _("Signatures");
	buf[T_URL] = _("URL");
	buf[T_VALIDATED_BY] = _("Validated By");
	buf[T_VERSION] = _("Version");

	for(i = 0; i < ARRAYSIZE(wbuf); i++) {
		wlen[i] = mbstowcs(wbuf[i], buf[i], strlen(buf[i]) + 1);
		wcol[i] = wcswidth(wbuf[i], wlen[i]);
		if(wcol[i] > maxcol) {
			maxcol = wcol[i];
		}
		if(wlen[i] > maxlen) {
			maxlen = wlen[i];
		}
	}

	for(i = 0; i < ARRAYSIZE(wbuf); i++) {
		size_t padlen = maxcol - wcol[i];
		wmemset(wbuf[i] + wlen[i], L' ', padlen);
		wmemcpy(wbuf[i] + wlen[i] + padlen, title_suffix, ARRAYSIZE(title_suffix));
		wcstombs(titles[i], wbuf[i], sizeof(wbuf[i]));
	}
}

```

rmarquis commented:
I quickly looked at the current implementation, and there are many issues left:

* hardcoding of language dependent value
* installed status and color missing (like cower, that I find much more readable than pacman colored output)
* requires capital to the Out Of Date argument (Yes/No)
* disturbing "roll out" effect when passing many arguments / need speed optimization
* breaks search

Also, please try to separate translation commits from code change commits.

ChuckDaniels87 commented:
>* hardcoding of language dependent value
I'll try to find how it is done upstream, but it seems hardcoded too.
>* installed status and color missing (like cower, that I find much more readable than pacman colored output)
>* requires capital to the Out Of Date argument (Yes/No)
I think these are easy to do.
>* disturbing "roll out" effect when passing many arguments / need speed optimization
I'll take a look on this
>* breaks search
I didn't noticed that was the same function. I can take out the code to a new function or maybe use a conditional inside SearchInfoAur.

rmarquis commented:
For the last one, simply split SearchInfoAur() into SearchAur() and InfoAur().

ChuckDaniels87 commented:
>- [ ] hardcoding of language dependent value
>- [x] ~~installed status and color missing (like cower, that I find much more readable than pacman colored output)~~
>- [x] ~~requires capital to the Out Of Date argument (Yes/No)~~
>- [ ] disturbing "roll out" effect when passing many arguments / need speed optimization
>- [x]  ~~breaks search~~

I think that roll out effect can be fixed buffering the info in a **tmp file**. What do you think? 
I'll try to optimize the code anyways.

rmarquis commented:
> I think that roll out effect can be fixed buffering the info in a tmp file. What do you think? 

Not really. You'd better try to get all the variables set for each packages, and then print the whole info in one go.

ChuckDaniels87 commented:
> Not really. You'd better try to get all the variables set for each packages, and then print the whole info in one go.
I found a solution storing all info into an array all the formatted text and print the array at the end of the function.

ChuckDaniels87 commented:
I've benchmarked both `cower` direct query and native `pacaur` formatting:

| Packages | `cower` (in secs) | `pacaur` (in secs) |
|:------------:|:---------:|:----------:|
|  1  | 0.30 | 0.48 |
|  2 |  0.31 | 0.65 |
|  4 | 0.33 | 0.99 |
|  8 | 0.37 | 1.67 |
|  16 | 0.41 | 3.00|

Formatting in `pacaur` lets translate the text but the performance drops in large queries. I think it can be optimized but it will be always slower than C code.

rmarquis commented:
I was expecting the code to be slower, but I am a bit surprised by the current performance. This said, the code of the PR is much longer than I initially thought it would be. Any idea what is the bottleneck here? I'd have expected expac to slow the operation, but it is called a single time here.

ChuckDaniels87 commented:
I've profiled pacaur with those changes but I can't see any real bottleneck ([check log here](https://gist.github.com/ChuckDaniels87/fad5ce08c3639a4d99422a0b1cce35fc)). Note that runs slower during the profiling.

Maybe the code could be reduced several lines but I aimed to make it readable/efficient. It is needed to take many considerations to properly format the text.

Nevertheless, if you think I should change something just tell me. Thanks for your quick feedback. :+1:

rmarquis commented:
While reviewing this code, I couldn't help myself but wonder why some part of it are so complicated. I used a very old WIP commit of mine and completed it using your code, and eventually reimplemented this feature in an alternative manner. More importantly, it runs faster too:

| AUR Packages | `cower` (in secs) | `pacaur` (in secs) |
|:------------:|:---------:|:----------:|
|  1  | 0.17 | 0.25 |
|  2 |  0.18 | 0.29 |
|  4 | 0.20 | 0.40 |
|  8 | 0.22 | 0.43 |
|  16 | 0.26 | 0.60|

My connection seems is bit better than yours, but I still observe a 3x - 4x speed improvement with large queries compared to your code. I did make heavy use of your work, so thanks for that! 

Current code uses 2 expac calls, so there is room for improvement. I did not include any translation, so you might want to send a new PR later (note there are a bunch of new translation strings coming soon).

ChuckDaniels87 commented:
I'll update Spanish translation soon, when you are done adding new strings. Try to test the code in several languages because special characters seem to cause troubles in formatting. I'm glad my work got use in pacaur! :+1:

ChuckDaniels87 commented:
I've checked your last commit but it seems broken when you use internationalization (Spanish):
```
:: Package pacaur not found in repositories, trying AUR...
Repositorio              : aur
Nombre                   : pacaur [installed: 4.6.9-1]
Version                  : 4.6.10-1
Descripción             : An AUR helper that minimizes user interaction
URL                      : https://github.com/rmarquis/pacaur
Página AUR              : https://aur.archlinux.org/packages/pacaur
Licencias                : ISC  
Keywords                 : AUR  cower  helper  wrapper  
Provee                   : Nada
Depende de               : cower  expac  sudo  git  
Dependencias Make        : perl  
Dependencias opcionales  : Nada
En conflicto con         : Nada
Reemplaza a              : Nada
Maintainer               : Spyhawk
Popularidad              : 59,54
Votos                    : 638
Obsoleto                 : No
Creado                   : mar 17 may 2011 23:16:03 CEST
Última modificación    : sáb 01 oct 2016 12:19:07 CEST

```
I didn't use `printf` approach I've seen in the rest of your code because cause internationalization issues.

Additionally:
* [installed: <version>] must be red to match cower output.
* Optional dependencies are not properly formatted, they are all in the same line. Cower and pacman format them one each line.
* Long description is not well formatted when you need to break the line.

rmarquis commented:
Are you _sure_ this isn't an issue with your translation file/format? Printf is used pretty much in all translated strings, and I'd be surprised if it only breaks for you here. I did see it breaking when translators use some weird characters or encoding though.

> 1. [installed: <version>] must be red to match cower output.

The color is red or green depending on the current version (older, or newer in the case of vcs packages). This is something you actually missed in the PR.

> 2. Optional dependencies are not properly formatted, they are all in the same line.
> 3. Long description is not well formatted when you need to break the line.

Yep, needs to fix that.

ChuckDaniels87 commented:
> Are you sure this isn't an issue with your translation file/format? Printf is used pretty much in all translated strings, and I'd be surprised if it only breaks for you here. I did see it breaking when translators use some weird characters or encoding though.

I'll regenerate the translation files (.po) with your code and I will PR so you can try.

> The color is red or green depending on the current version (older, or newer in the case of vcs packages). This is something you actually missed in the PR.

The version is shown when version installed is not the latest version in AUR and it is coloured red. This colouring does not depends on Out of Date flag:
![captura de pantalla de 2016-10-04 12-50-13](https://cloud.githubusercontent.com/assets/8478202/19071720/c8f00322-8a31-11e6-8da6-94c6c734fedd.png)

rmarquis commented:
I have some trouble making foreign locale work for some reason (and never took time to fix it), so I'm afraid I won't be able to test it. But now, I believe that is the `GetLength` function that doesn't count 'weird' characters for some reason.

Also, I inverted the color in the version status check (should be green is version is newer, red otherwise).

ChuckDaniels87 commented:
> I have some trouble making foreign locale work for some reason (and never took time to fix it), so I'm afraid I won't be able to test it. 

I've tested recently and i seems to work. Remember that in Arch you need to generate the [locales first](https://wiki.archlinux.org/index.php/Locale#Generating_locales). You need to set `LANG=` to the full locale name. For example: `LANG=es_ES.UTF-8 pacaur`

> But now, I believe that is the GetLength function that doesn't count 'weird' characters for some reason.

GetLength() seems to work well. At least, I used it to detect the longest string and it get the characters.

```bash
$> echo "English: $(GetLength "versión")" && echo "Spanish: $(GetLength "version")"
English: 7
Spanish: 7
```

rmarquis commented:
> I've tested recently and i seems to work. Remember that in Arch you need to generate the locales first. You need to set LANG= to the full locale name. For example: LANG=es_ES.UTF-8 pacaur

Yes, doesn't work, even after a reboot. The es_ES.UTF-8 locale is correctly listed in `locale -a`. I have been aware of that issue for a long time but I simply didn't care enough to fix it.

What's weird is that I tried to replace "Last submitted" by "Última modificación" directly in the printf code and it indeed works as expected.

```
$ pacaur -Si pacaur-git
:: Package pacaur-git not found in repositories, trying AUR...
Repository           : aur
Name                 : pacaur-git [installed: 4.6.10.28.g1cd2740-1]
Version              : 4.6.10-1
Description          : An AUR helper that minimizes user interaction
URL                  : https://github.com/rmarquis/pacaur
AUR Page             : https://aur.archlinux.org/packages/pacaur-git
Licenses             : ISC
Keywords             : AUR  cower  helper  wrapper
Provides             : pacaur
Depends on           : cower  expac  sudo  git
Make Deps            : perl
Optional Deps        : None
Conflicts With       : pacaur
Replaces             : None
Maintainer           : Spyhawk
Popularity           : 0,43
Votes                : 25
Out of Date          : No
Submitted            : Wed Nov 16 20:20:14 2011
Última modificación  : Sat Oct  1 12:19:18 2016
```

ChuckDaniels87 commented:
I don't know what it is happening but I tried the same as you did and this is the result:

```
./pacaur -i pacaur
Repository             : aur
Name                   : pacaur [installed: 4.6.9-1]
Version                : 4.6.10-1
Description            : An AUR helper that minimizes user interaction
URL                    : https://github.com/rmarquis/pacaur
AUR Page               : https://aur.archlinux.org/packages/pacaur
Licenses               : ISC  
Keywords               : AUR  cower  helper  wrapper  
Provides               : None
Depends on             : cower  expac  sudo  git  
Make Deps              : perl  
Optional Deps          : None
Conflicts With         : None
Replaces               : None
Maintainer             : Spyhawk
Popularity             : 60,44
Votes                  : 639
Out of Date            : No
Submitted              : Tue May 17 23:16:03 2011
Última modificación  : Sat Oct  1 12:19:07 2016
```

rmarquis commented:
Specifically, what I did is replace `$"Last Modified"` by `$"Última modificación"` in the `infolabel` array.

Maybe a prompt setting? I'm using standard bash with konsole/yakuake.

ChuckDaniels87 commented:
> Specifically, what I did is replace $"Last Modified" by $"Última modificación" in the infolabel array.

I did the same.

> Maybe a prompt setting? I'm using standard bash with konsole/yakuake.

I tried gnome-terminal and pure TTY with both bash and zsh.

Many stuff of my PR was about trying to fix this issue messing up everything. It seems something in my system is doing something really stupid:

```bash
bash -c "echo $LANG; printf '%-20s|\n%-20s|\n' 'versión' 'version'"
es_ES.utf8
versión            |
version             |
```

I found an [old Debian issue] about this. I'll keep looking for something related to this bug but I think your code will be fine after all.

ChuckDaniels87 commented:
This is pretty strange but I think it will definitely impact `pacaur` formatting. To completely discard that it is my configuration issue, would you mind to try in [this website](http://www.learnshell.org/) this code:
```
printf '%-20s|\n%-20s|\n' 'versión' 'version'
```
My output is:
```
versión            |
version             |
```

ChuckDaniels87 commented:
Well, I found a clarifying [thread in GNU coreutils](https://lists.gnu.org/archive/html/bug-coreutils/2014-04/msg00049.html) mailing lists. Hence, it is a real problem using `printf`.

rmarquis commented:
> Hence, it is a real problem using printf.

It might work correctly on some version only. It does work on my machine, and I haven't any specific setup.

ChuckDaniels87 commented:
[The answer](http://lists.gnu.org/archive/html/bug-bash/2014-04/msg00022.html) of a Bash developer is pretty clear about `printf`.

rmarquis commented:
Should be fixed in f3d89beb3c91d95071cc8c7ff5f6ac551c699844.

ChuckDaniels87 commented:
> Edit2: I don't get this. When my locale is set to en_US.UTF-8, the output is correct (unlike the ML example). It's buggy only with some locales.

It is just a mess and totally random.

> Should be fixed in f3d89be.

I think it is not because every string could have different number of Unicode characters. GetLength() is working good, that function is not the problem.

```
:: Paquete(s) pacaur no se encontraron en los repositorios, intentando AUR...
Repositorio              : aur
Nombre                   : pacaur [instalado]
Versión                 : 4.6.10-1
Descripción             : An AUR helper that minimizes user interaction
URL                      : https://github.com/rmarquis/pacaur
Página AUR              : https://aur.archlinux.org/packages/pacaur
Licencias                : ISC  
Palabras clave           : AUR  cower  helper  wrapper  
Provee                   : Nada
Depende de               : cower  expac  sudo  git  
Dependencias make        : perl  
Dependencias opcionales  : Nada
En conflicto con         : Nada
Reemplaza a              : Nada
Encargado                : Spyhawk
Popularidad              : 60,14
Votos                    : 639
Obsoleto                 : No
Fecha de creación       : mar 17 may 2011 23:16:03 CEST
Última modificación    : sáb 01 oct 2016 12:19:07 CEST

```

rmarquis commented:
Guess that fixed some of the case only... Damn, the unicode offset has to be computed for each line individually. I now understand why your code was so slow.

To be honest, I feel like I simply won't fix that. From my pov, a better performance in my daily use is more important than some display issue for internationalization that I simply don't care about. Maybe I should even just drop i18n entirely, and be done with it.

ChuckDaniels87 commented:
Don't give up, we almost got it! I'm going to try one thing I used in my PR and profile the code.

rmarquis commented:
Yeah. But not tonight, I need a good night of sleep.

Should be possible to simply compute the offset between characters and unicode bytes, but that idiot shell don't see any difference sometimes.

ChuckDaniels87 commented:
Ok, I just did it. This solution does not affect if you increase the number of packages. In fact, it has no noticeable impact in the speed of the query.

rmarquis commented:
Note on the Optional Deps: I've noticed it's not perfect when maxlength is reached, but it is actually similar to cower implementation. Tried to fix it properly but I didn't bother very long.

I'll see for the roll out effect. Haven't tested your change yet, but roll out effect was practiclaly unnoticeable on my machine.

ChuckDaniels87 commented:
> Note on the Optional Deps: I've noticed it's not perfect when maxlength is reached, but it is actually similar to cower implementation. Tried to fix it properly but I didn't bother very long.

I've fixed not only long lines breaking but also proper regex to catch packages. I've implemented in only one line with two `sed` calls and a `fold` call (#570), so it will be fast.

> I'll see for the roll out effect. Haven't tested your change yet, but roll out effect was practiclaly unnoticeable on my machine.

I can notice it in slow machines (p.e. my laptop). The temp file solution is really efficient to get rid of it.

rmarquis commented:
> I can notice it in slow machines (p.e. my laptop). The temp file solution is really efficient to get rid of it.

Just being curious, what's your laptop spec? Mine is also a quite old laptop (6y, standard HD 5400rpm) and it wasn't high end from the start either.

ChuckDaniels87 commented:
Sony Vaio Intel Core i5 from 2010 with ATI Radeon graphics (using open source driver). I've replaced my HDD with a SDD recently.

rmarquis commented:
Interesting. I'd have thought using a SSD would really improve display speed.

I'm not too fond of using temporary file (since I reworked the work a while ago to avoid them entirely, apart from lock files), but this is now pushed in 84e3e6edb3dc33d8a27c9dbf3236a4f7c3f4a7ab.

ChuckDaniels87 commented:
> Interesting. I'd have thought using a SSD would really improve display speed.

Not really, you are not writting the info in the disk.

> I'm not too fond of using temporary file (since I reworked the work a while ago to avoid them entirely, apart from lock files), but this is now pushed in 84e3e6e. 

I think with the last commit this new internationalization feature is complete!  It was tricky one but fun. :+1: 

I'll PR new Spanish translation when you are done adding new stuff.
